riscv-sbi.adoc

```diff
@@ -257,6 +258,93 @@ call doesn't return.
 | *RESERVED*                |             |    0x09-0x0F |
 |===
 
+== Hart State Managment Extension, Extension ID: 0x48534D (HSM)
+
+The Hart State Management Extension introduces a set of functions that allow the
+supervisor software to request higher privilege mode to start/stop harts running
+in supervisor mode.
+
+[source, C]
+----
+struct sbiret sbi_hart_start(unsigned long hartid, unsigned long start_addr, unsigned
+long priv)
+----
+
+Brings up *hartid* either during initial boot or after an `sbi_hart_stop` SBI
+call in supervisor mode. This will be an asynchronous call. That means the caller
+on the supervisor side needs to verify if the hart is really started or not.
+
+*start_addr* points to a runtime-specified physical address, where a hart can
+resume execution after its initialization/resume sequence.  Before jumping to
+*start_addr*, the hart MUST configure PMP if present and switch to Supervisor
+mode.
+
+*priv* is an opaque pointer to a opaque data segment that the caller can use to pass
+information about execution context.
+
+*Returns* one of the following possible SBI error codes thorugh sbiret.error.
+
+[cols="<,>",options="header,compact"]
+|===
+| Error code                | Description
+| SBI_SUCCESS               | Hart was previously in stopped state. It will start executing from `start_addr`.
+| SBI_ERR_INVALID_ADDRESS   | `start_addr` is not valid possibly due to following reasons. +
+                              * it is not a valid physical address. +
+                              * The address is prohibited by PMP to run in supervisor mode +
+| SBI_ERR_INVALID_PARAM     | `hartid` is not a valid hartid as corresponding hart cannot started in supervisor mode.
+| SBI_ERR_ALREADY_AVAILABLE | The given hartid is already started.
+| SBI_ERR_FAILED            | The start request failed for unknown reasons.
+|===
+
+[source, C]
+----
+struct sbiret sbi_hart_stop()
+----
+
+Stop the calling hart by removing it from the list of harts that are currently running
+supervisor mode software. The stopped hart will no longer execute supervisor or user mode
+code, and will not respond to inbound supervisor or user IPIs.
+A hart can only stop itself and can be only started again via the `sbi_hart_start`
+call. This is a synchronous call and is not expected to return if succeeds.
+
+*Returns* following SBI error code through sbiret.error only if it fails.
+
+* SBI_ERR_FAILED
```

Comment:
Why would this call fail, other than SBI_ERR_NOT_SUPPORTED?

Reply:
This is added for future compatibility. Currently, stop() will just call WFI. But we may have some actual hardware calls in future that will perform more than WFI. If those calls fail, machine mode can return SBI_ERR_FAILED.